docs/example_java.md

```diff
@@ -3,9 +3,314 @@ title: "Example: Java"
+    <img src="assets/docimg/example-java-image-1.png">
+  </a>
+  <figcaption>The server is up! Click the screenshot to see an interactive snapshot.</figcaption>
+</figure>
+
+## Step 1: Let's Add Benchmark Trickery
+
+Before we try to make this faster, let's measure it.
+
+In addition to running things in your cluster, Tilt can run commands locally.
+You can direct Tilt to execute existing scripts or arbitrary shell commands on
+your own machine.
+
+We want to measure the time from code change to new process. To do that, we'll
+use [`local_resource`](local_resource.html), which lets you locally
+run scripts, shell code, or servers, and manage them from your sidebar like any
+other Tilt resource.
+
+First, we add a `local_resource` to our
+[Tiltfile](https://github.com/windmilleng/tilt-example-java/blob/master/1-measured/Tiltfile)
+that records the start time in a Java file.
+
+```python
+k8s_resource(
+    'example-java', 
+    port_forwards=8000, 
+    resource_deps=['deploy'])
+
+# Records the current time, then kicks off a server update.
+# Normally, you would let Tilt do deploys automatically, but this
+# shows you how to set up a custom workflow that measures it.
+local_resource(
+    'deploy',
+    './record-start-time.sh',
+)
+```
+
+The `local_resource()` call creates a local resource named `deploy`. The second
+argument is the script that it runs.
+
+We've also modified our server to read that start time, calculate the time
+elapsed, then display this in both logs and HTML.
+
+Let's click the button on the `deploy` resource and see what happens!
+
+<figure>
+  <a class="is-image" href="https://cloud.tilt.dev/snapshot/AeDqiucLr-00XLJpiWc=">
+    <img src="assets/docimg/example-java-image-2.png">
+  </a>
+  <figcaption>Step 1 complete. Click the screenshot to see an interactive snapshot.</figcaption>
+</figure>
+
+| Approach | Deploy Time |
+|---|---|
+| Naive | 87.7s |
+{:.benchmark-report}
+
+If you look closely, the elapsed time displayed in the Tilt sidebar is different
+than the benchmark our app logged. That's OK! In multi-service development,
+there are many benchmarks we care about -- the time to build the image, the time
+to schedule the process, and the time until the server is ready to serve
+traffic. 
+
+The Tilt sidebar gives you some default benchmarks, _and_ the tools to capture
+your own benchmarks.
+
+Our benchmarks show this is slow. Can we do better?
+
+## Step 2: Let's Optimize for the Java Toolchain
+
+What's taking up so much time? The logs show that when we make the change to a file, we:
+
+1) Copy the source files to the image.
+
+2) Download Gradle.
+
+3) Download all the Spring dependencies.
+
+4) Compile the Java jar from scratch.
+
+But the Java community has done a lot of work to make caching dependencies and
+incremental compiles fast.  How can we better use the tools how they're meant
+to be used?
+
+With `local_resource`, we can compile the executable Jar locally, and copy it
```

Comment:
some users will want to compile the executable in the container for uniformity reasons--so maybe worth noting something like "If you feel strongly about compiling the executable inside the container , e.g. to maintain uniformity across your team, you can still shave off a lot of time here with Dockerfile optimizations, see [some blog post on the internet about this]".